RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc

```diff
@@ -0,0 +1,285 @@
+
+      AlgebraicSymMatrix66 covFinalReco;
+      GlobalVector p3FinalReco_glob, r3FinalReco_globv;
+      getFromFTS(finalrecostate, p3FinalReco_glob, r3FinalReco_globv, chargeReco, covFinalReco);
+
+
+      //To transform the global propagated track to local coordinates
+      int SegmentNumber = 0;
+
+      reco::ME0Muon MuonCandidate;
+      double ClosestDelR = 999.;
+
+      for (auto thisSegment = OurSegments->begin(); thisSegment != OurSegments->end(); 
+	   ++thisSegment,++SegmentNumber){
+	ME0DetId id = thisSegment->me0DetId();
+
+	auto roll = me0Geom->etaPartition(id); 
+
+	if ( zSign * roll->toGlobal(thisSegment->localPosition()).z() < 0 ) continue;
+
+	GlobalPoint r3FinalReco_glob(r3FinalReco_globv.x(),r3FinalReco_globv.y(),r3FinalReco_globv.z());
+
+	LocalPoint r3FinalReco = roll->toLocal(r3FinalReco_glob);
+	LocalVector p3FinalReco=roll->toLocal(p3FinalReco_glob);
+
+	LocalPoint thisPosition(thisSegment->localPosition());
+	LocalVector thisDirection(thisSegment->localDirection().x(),thisSegment->localDirection().y(),thisSegment->localDirection().z());  //FIXME
+
+	//The same goes for the error
+	AlgebraicMatrix thisCov(4,4,0);   
+	for (int i = 1; i <=4; i++){
+	  for (int j = 1; j <=4; j++){
+	    thisCov(i,j) = thisSegment->parametersError()(i,j);
+	  }
+	}
+
+	/////////////////////////////////////////////////////////////////////////////////////////
+
+
+	LocalTrajectoryParameters ltp(r3FinalReco,p3FinalReco,chargeReco);
+	JacobianCartesianToLocal jctl(roll->surface(),ltp);
+	AlgebraicMatrix56 jacobGlbToLoc = jctl.jacobian(); 
+
+	AlgebraicMatrix55 Ctmp =  (jacobGlbToLoc * covFinalReco) * ROOT::Math::Transpose(jacobGlbToLoc); 
+	AlgebraicSymMatrix55 C;  // I couldn't find any other way, so I resort to the brute force
+	for(int i=0; i<5; ++i) {
+	  for(int j=0; j<5; ++j) {
+	    C[i][j] = Ctmp[i][j]; 
+
+	  }
+	}  
+
+	Double_t sigmax = sqrt(C[3][3]+thisSegment->localPositionError().xx() );      
+	Double_t sigmay = sqrt(C[4][4]+thisSegment->localPositionError().yy() );
+
+	bool X_MatchFound = false, Y_MatchFound = false, Dir_MatchFound = false;
+	
+
+	 // if ( (fabs(thisPosition.x()-r3FinalReco.x()) < (3.0 * sigmax)) || (fabs(thisPosition.x()-r3FinalReco.x()) < 2.0 ) ) X_MatchFound = true;
+	 // if ( (fabs(thisPosition.y()-r3FinalReco.y()) < (3.0 * sigmay)) || (fabs(thisPosition.y()-r3FinalReco.y()) < 2.0 ) ) Y_MatchFound = true;
+
+	 // if ( fabs(p3FinalReco_glob.phi()-roll->toGlobal(thisSegment->localDirection()).phi()) < 0.15) Dir_MatchFound = true;
+
+
+	 if ( (fabs(thisPosition.x()-r3FinalReco.x()) < (X_PULL_CUT * sigmax)) || (fabs(thisPosition.x()-r3FinalReco.x()) < X_RESIDUAL_CUT ) ) X_MatchFound = true;
+	 if ( (fabs(thisPosition.y()-r3FinalReco.y()) < (Y_PULL_CUT * sigmay)) || (fabs(thisPosition.y()-r3FinalReco.y()) < Y_RESIDUAL_CUT ) ) Y_MatchFound = true;
+
+	 if ( fabs(p3FinalReco_glob.phi()-roll->toGlobal(thisSegment->localDirection()).phi()) < PHIDIR_RESIDUAL_CUT) Dir_MatchFound = true;
```

Comment:
did you check that this does 2*pi wraparound correctly?
otherwise, use reco::deltaPhi